x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java

```diff
@@ -0,0 +1,551 @@
+        }
+
+        @Override
+        public SortField sortField(Object missingValue, MultiValueMode sortMode, Nested nested, boolean reverse) {
+            XFieldComparatorSource source = new BytesRefFieldComparatorSource(this, missingValue,
+                    sortMode, nested);
+            return new SortField(getFieldName(), source, reverse);
+        }
+    
+    }
+
+    private int ignoreAbove;
+
+    private WildcardFieldMapper(String simpleName, MappedFieldType fieldType, MappedFieldType defaultFieldType,
+                int ignoreAbove, Settings indexSettings, MultiFields multiFields, CopyTo copyTo) {
+        super(simpleName, fieldType, defaultFieldType, indexSettings, multiFields, copyTo);
+        this.ignoreAbove = ignoreAbove;
+        assert fieldType.indexOptions() == IndexOptions.DOCS;
+        
+        ngramFieldType = new FieldType();
+        ngramFieldType.setTokenized(true);            
+        ngramFieldType.setIndexOptions(IndexOptions.DOCS);
+        ngramFieldType.setOmitNorms(true);
+        ngramFieldType.freeze();
+    }
+
+
+    /** Values that have more chars than the return value of this method will
+     *  be skipped at parsing time. */
+    // pkg-private for testing
+    int ignoreAbove() {
+        return ignoreAbove;
+    }    
+    
+    @Override
+    protected WildcardFieldMapper clone() {
+        return (WildcardFieldMapper) super.clone();
+    }
+
+    @Override
+    public WildcardFieldType fieldType() {
+        return (WildcardFieldType) super.fieldType();
+    }
+
+    @Override
+    protected void doXContentBody(XContentBuilder builder, boolean includeDefaults, Params params) throws IOException {
+        super.doXContentBody(builder, includeDefaults, params);
+        if (includeDefaults || ignoreAbove != Defaults.IGNORE_ABOVE) {
+            builder.field("ignore_above", ignoreAbove);
+        }
+    }
+    
+    @Override
+    protected void parseCreateField(ParseContext context, List<IndexableField> fields) throws IOException {
+        final String value;
+        if (context.externalValueSet()) {
+            value = context.externalValue().toString();
+        } else {
+            XContentParser parser = context.parser();
+            if (parser.currentToken() == XContentParser.Token.VALUE_NULL) {
+                value = fieldType().nullValueAsString();
+            } else {
+                value =  parser.textOrNull();
+            }
+        }
+        ParseContext.Document parseDoc = context.doc();
+        
+        createFields(value, parseDoc, fields);        
+    }   
+    
+    // For internal use by Lucene only - used to define ngram index
+    final FieldType ngramFieldType;
+    
+    void createFields(String value, Document parseDoc, List<IndexableField>fields) {
+        if (value == null || value.length() > ignoreAbove) {
+            return;
+        }
+        NGramTokenizer tokenizer = new NGramTokenizer(NGRAM_SIZE, NGRAM_SIZE);
```

Comment:
can you create an analyzer in order to reuse attributes, which tend to be costly to instantiate? I'm thinking of creating an analyzer in the constructor, e.g.

```
ngramAnalyzer = new Analyzer() {
    @Override
    public TokenStreamComponents createComponents(String fieldName) {
      Tokenizer tokenizer = new NGramTokenizer(NGRAM_SIZE, NGRAM_SIZE);
      return new TokenStreamComponents(tokenizer);
    }
};
```
And then calling `ngramAnalyzer#analyze` here instead of creating the tokenizer manually.

Reply:
I may be missing something but I'm not sure that works. Calling `Analyzer.tokenStream(name,reader)` reuses the same NgramTokenizer instance and I'm now getting errors about reset/close contract violations. In the case of JSON docs with arrays of values for a field the analyzer reuses the same NGramTokenizer which has `setReader` called on it with different values before the doc is actually indexed and the token streams consumed.

Reply:
I don't understand why it would break the reset/close contract to use a single analyzer built in the constructor. Can you show the error that you're getting ?